apps/prairielearn/src/ee/lib/billing/plan-grants.ts

```diff
@@ -0,0 +1,47 @@
+import { CourseInstanceSchema, InstitutionSchema } from '../../../lib/db-types';
+import { getEnrollmentForUserInCourseInstance } from '../../../models/enrollment';
+import {
+  getPlanGrantsForContext,
+  getPlanNamesFromPlanGrants,
+  getRequiredPlansForCourseInstance,
+} from '../../lib/billing/plans';
+import { planGrantsMatchPlanFeatures } from './plans-types';
+
+export async function checkPlanGrants(res) {
+  // We'll only check plan grants for course instances, as students can't
+  // currently ever access a course directly. And even if they could, plan
+  // grans aren't associated with courses.
+  if (!res.locals.course_instance) {
+    return true;
+  }
+
+  // We won't check plan grants if the user has a specific role in the course
+  // or course instance. We always grant instructors access to data.
+  //
+  // TODO: is this conditional checking the correct values?
+  // TODO: how do we account for user overrides when checking plan grants?
```

Comment:
If it's possible to do the check based on a user override then I think that would be nice because it would allow instructors to emulate a user to see whether they have access, and so see what the "upgrade" page looks like. If we only check the authn user then an instructor emulating a student user will still have access granted, which I think would be confusing? I say all this under the assumption that it would be easy to use the current user override rather than the authenticated user. If this is hard for some reason then it's definitely not a hard requirement to do the check based on the user override.

Reply:
I guess that's reasonable, yeah. I imagine it'd still be nice to have a break-glass option for the instructor, but I guess we can hook into the "Student view without access restrictions" machinery via the `pl_requested_course_role`/`pl_requested_course_instance_role` cookies (or whatever data on `res.locals.authz_data` that turns into).

Reply:
I made this change in d982196 / 4e926a1. `res.locals.authz_data.course_role` and `res.locals.authz_data.course_instance_role` already reflect the role override headers, so things generally just worked without much fussing!